crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,532 @@
+                        "LineWithItems::new() should uphold the invariant that this list is always non-empty"
+                    )
+                };
+                let range = this_range.map_or(*first_range, |r| {
+                    TextRange::new(r.start(), first_range.end())
+                });
+                all_items.push(DunderAllItem {
+                    value: first_val.clone(),
+                    original_index: all_items.len(),
+                    range,
+                    additional_comments: *comment_range,
+                });
+                this_range = None;
+                for (value, range) in rest {
+                    all_items.push(DunderAllItem {
+                        value: value.clone(),
+                        original_index: all_items.len(),
+                        range: *range,
+                        additional_comments: None,
+                    });
+                }
+            }
+        }
+    }
+    all_items
+}
+
+#[derive(Clone, Debug)]
+struct DunderAllItem {
+    value: String,
+    // Each `AllItem` in any given list should have a unique `original_index`:
+    original_index: usize,
+    // Note that this range might include comments, etc.
+    range: TextRange,
+    additional_comments: Option<TextRange>,
+}
+
+impl Ranged for DunderAllItem {
+    fn range(&self) -> TextRange {
+        self.range
+    }
+}
+
+impl PartialEq for DunderAllItem {
+    fn eq(&self, other: &Self) -> bool {
+        self.original_index == other.original_index
+    }
+}
+
+impl Eq for DunderAllItem {}
+
+impl DunderAllItem {
+    fn sort_index(&self) -> (&str, usize) {
+        (&self.value, self.original_index)
+    }
+}
+
+impl Ord for DunderAllItem {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.sort_index().cmp(&other.sort_index())
+    }
+}
+
+impl PartialOrd for DunderAllItem {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+fn join_singleline_dunder_all_items(sorted_items: &[DunderAllItem], locator: &Locator) -> String {
+    sorted_items
+        .iter()
+        .map(|item| locator.slice(item))
+        .join(", ")
+}
+
+fn join_multiline_dunder_all_items(
+    sorted_items: &[DunderAllItem],
+    locator: &Locator,
+    parent: &ast::Stmt,
+    additional_indent: &str,
+    newline: &str,
+    needs_trailing_comma: bool,
+) -> Option<String> {
+    let indent = indentation(locator, parent)?;
+    let mut new_dunder_all = String::new();
+    for (i, item) in sorted_items.iter().enumerate() {
+        new_dunder_all.push_str(indent);
+        new_dunder_all.push_str(additional_indent);
+        new_dunder_all.push_str(locator.slice(item));
+        let is_final_item = i == (sorted_items.len() - 1);
+        if !is_final_item || needs_trailing_comma {
+            new_dunder_all.push(',');
+        }
+        if let Some(comment) = item.additional_comments {
+            new_dunder_all.push_str("  ");
```

Comment:
This might be a bit opinionated for some :laughing:. Any chance we can preserve the original spacing?

Reply:
Ew @ anybody who doesn't put two spaces before an inline comment, but: see 655bf67ae3459da6b0bec34ab959d56cd4d26bfe :)